File: packages/core/src/programmers/llm/LlmMetadataFactory.ts
import { ILlmSchema } from "@typia/interface";
import ts from "typescript";

import { ITypiaContext } from "../../context/ITypiaContext";
import { TransformerError } from "../../context/TransformerError";
import { MetadataFactory } from "../../factories/MetadataFactory";
import { MetadataStorage } from "../../factories/MetadataStorage";
import { MetadataObject } from "../../schemas/metadata/MetadataObject";
import { MetadataSchema } from "../../schemas/metadata/MetadataSchema";
import { ValidationPipe } from "../../typings/ValidationPipe";

export namespace LlmMetadataFactory {
  export const getConfig = (props: {
    context: ITypiaContext;
    method: string;
    node: ts.TypeNode | undefined;
  }):
    | Partial<
        ILlmSchema.IConfig & {
          equals: boolean;
        }
      >
    | undefined => {
    if (props.node === undefined) return undefined;

    const type: ts.Type = props.context.checker.getTypeFromTypeNode(props.node);
    const collection: MetadataStorage = new MetadataStorage();
    const result: ValidationPipe<MetadataSchema, MetadataFactory.IError> =
      MetadataFactory.analyze({
        checker: props.context.checker,
        transformer: props.context.transformer,
        options: {
          absorb: true,
          escape: false,
          constant: true,
          functional: false,
        },
        components: collection,
        type,
      });
    if (result.success === false)
      throw new TransformerError({
        code: `typia.llm.${props.method}`,
        message: `Failed to analyze generic argument "Config".`,
      });

    const meta: MetadataSchema = result.data;
    if (
      meta.size() !== 1 ||
      meta.objects.length !== 1 ||
      meta.nullable === true ||
      meta.isRequired() === false
    )
      throw new TransformerError({
        code: `typia.llm.${props.method}`,
        message: `Invalid generic argument "Config". It must be a literal object type.`,
      });

    const obj: MetadataObject = meta.objects[0]!;
    if (obj.type.properties.some((p) => p.key.isSoleLiteral() === false))
      throw new TransformerError({
        code: `typia.llm.${props.method}`,
        message: `Invalid generic argument "Config". It must be a literal object type. Do not allow dynamic properties.`,
      });
    else if (
      obj.type.properties.some(
        (p) =>
          p.value.size() !== 1 ||
          p.value.constants.length !== 1 ||
          p.value.nullable === true ||
          p.value.isRequired() === false,
      )
    )
      throw new TransformerError({
        code: `typia.llm.${props.method}`,
        message: `Invalid generic argument "Config". It must be a literal object type. Do not allow variable type.`,
      });
    const config: Partial<ILlmSchema.IConfig> = {};
    for (const prop of obj.type.properties) {
      const key: string = prop.key.getSoleLiteral()!;
      const value: boolean | bigint | number | string =
        prop.value.constants[0]!.values[0]!.value;
      if (typeof value === "bigint")
        throw new TransformerError({
          code: `typia.llm.${props.method}`,
          message: `Invalid generic argument "Config". It must be a literal object type. Do not allow bigint.`,
        });
      (config as any)[key] = value;
    }
    return config;
  };
}

File: packages/typia/native/core/programmers/llm/LlmMetadataFactory.go
package llm

import (
  shimast "github.com/microsoft/typescript-go/shim/ast"
  nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
  nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
  nativemetadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

type llmMetadataFactoryNamespace struct{}

var LlmMetadataFactory = llmMetadataFactoryNamespace{}

func (llmMetadataFactoryNamespace) GetConfig(props struct {
  Context nativecontext.ITypiaContext
  Method  string
  Node    *shimast.Node
}) map[string]any {
  if props.Node == nil {
    return nil
  }

  typ := props.Context.Checker.GetTypeFromTypeNode(props.Node)
  collection := nativemetadata.NewMetadataCollection()
  result := nativefactories.MetadataFactory.Analyze(nativefactories.MetadataFactory_IProps{
    Checker:     props.Context.Checker,
    Transformer: props.Context.Transformer,
    Options: nativefactories.MetadataFactory_IOptions{
      Absorb:     true,
      Escape:     false,
      Constant:   true,
      Functional: false,
    },
    Components: collection,
    Type:       typ,
  })
  if result.Success == false {
    panic(nativecontext.NewTransformerError(nativecontext.TransformerError_IProps{
      Code:    "typia.llm." + props.Method,
      Message: "Failed to analyze generic argument \"Config\".",
    }))
  }

  meta := result.Data
  if meta.Size() != 1 ||
    len(meta.Objects) != 1 ||
    meta.Nullable == true ||
    meta.IsRequired() == false {
    panic(nativecontext.NewTransformerError(nativecontext.TransformerError_IProps{
      Code:    "typia.llm." + props.Method,
      Message: "Invalid generic argument \"Config\". It must be a literal object type.",
    }))
  }

  obj := meta.Objects[0]
  for _, p := range obj.Type.Properties {
    if p.Key.IsSoleLiteral() == false {
      panic(nativecontext.NewTransformerError(nativecontext.TransformerError_IProps{
        Code:    "typia.llm." + props.Method,
        Message: "Invalid generic argument \"Config\". It must be a literal object type. Do not allow dynamic properties.",
      }))
    }
  }
  for _, p := range obj.Type.Properties {
    if p.Value.Size() != 1 ||
      len(p.Value.Constants) != 1 ||
      p.Value.Nullable == true ||
      p.Value.IsRequired() == false {
      panic(nativecontext.NewTransformerError(nativecontext.TransformerError_IProps{
        Code:    "typia.llm." + props.Method,
        Message: "Invalid generic argument \"Config\". It must be a literal object type. Do not allow variable type.",
      }))
    }
  }

  config := map[string]any{}
  for _, prop := range obj.Type.Properties {
    key := prop.Key.GetSoleLiteral()
    constant := prop.Value.Constants[0]
    value := constant.Values[0].Value
    if constant.Type == "bigint" {
      panic(nativecontext.NewTransformerError(nativecontext.TransformerError_IProps{
        Code:    "typia.llm." + props.Method,
        Message: "Invalid generic argument \"Config\". It must be a literal object type. Do not allow bigint.",
      }))
    }
    config[*key] = value
  }
  return config
}
